tsdb/fileutil/direct_io_writer.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build linux
+
+package fileutil
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	// defaultAlignment is deliberately set higher to cover most setups. We rely on statx for more precise alignment.
```

Comment:
I didn't understand any of this.  Perhaps writing `statx(2)` would give more of a hint?

https://man7.org/linux/man-pages/man2/statx.2.html